### PR TITLE
[Openload.co] Fixed Extraction

### PR DESCRIPTION
**Pull request by sudovijay**
Just updated the code. i don't do much python still i tried to convert my code. lemme know if there is any prob with it

## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [ ] At least skimmed through [adding new extractor tutorial](https://github.com/rg3/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/rg3/youtube-dl#youtube-dl-coding-conventions) sections
- [ ] [Searched](https://github.com/rg3/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [ ] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Explanation of your *pull request* in arbitrary form goes here. Please make sure the description explains the purpose and effect of your *pull request* and is worded well enough to be understood. Provide as much context and examples as possible.

### CODE REVIEW

File: youtube_dl/extractor/openload.py [hunk 1]
```diff
@@ -75,17 +75,20 @@ def _real_extract(self, url):
...
+        num = 2
 
         while num < len(ol_id):
-            urlcode += compat_chr(int(float(ol_id[num:][:3])) +
-                                  first_three_chars - fifth_char * int(float(ol_id[num + 3:][:2])))
+            key = int(float(ol_id[num + 3:][:2]))
+            urlcode[key] = compat_chr(int(float(ol_id[num:][:3])) - first_two_chars)
             num += 5
+        
+        sorted(urlcode, key=lambda key: urlcode[key])
```
Comment (yan12125) [inline youtube_dl/extractor/openload.py]: What's the purpose of this line? ```sorted``` does not change the target
```
>>> a = [1, 3, 2]
>>> sorted(a)
[1, 2, 3]
>>> a
[1, 3, 2]
```

Comment (sudovijay) [inline youtube_dl/extractor/openload.py]: its for dictionaries. sorting by keys ! if you got some other way to do so just lemme know :)

Comment (yan12125) [inline youtube_dl/extractor/openload.py]: The point is that ```urlcode``` is not modified. I change the code to:
```
        print(urlcode)
        print(sorted(urlcode, key=lambda key: urlcode[key]))
        print(urlcode)
```
And it prints:
```
{30: '.', 19: '6', 39: 'a', 8: 'U', 41: 'e', 10: 'o', 32: '.', 35: 's', 5: 'c', 11: '~', 40: 'G', 23: '1', 34: '~', 42: 'F', 9: '9', 25: '0', 7: 's', 3: 'f', 27: '1', 37: '4', 20: '1', 14: '8', 1: 'U', 21: '9', 18: '3', 31: '0', 24: '4', 28: '1', 12: '1', 2: 'E', 16: '5', 4: 'G', 22: '~', 15: '6', 17: '6', 29: '2', 0: 'k', 13: '4', 38: 'h', 33: '0', 26: '.', 36: 'U', 6: 'l'}
[30, 32, 26, 25, 31, 33, 23, 27, 20, 28, 12, 29, 18, 37, 24, 13, 16, 19, 15, 17, 14, 9, 21, 2, 42, 40, 4, 8, 1, 36, 39, 5, 41, 3, 38, 0, 6, 10, 35, 7, 11, 34, 22]
{30: '.', 19: '6', 39: 'a', 8: 'U', 41: 'e', 10: 'o', 32: '.', 35: 's', 5: 'c', 11: '~', 40: 'G', 23: '1', 34: '~', 42: 'F', 9: '9', 25: '0', 7: 's', 3: 'f', 27: '1', 37: '4', 20: '1', 14: '8', 1: 'U', 21: '9', 18: '3', 31: '0', 24: '4', 28: '1', 12: '1', 2: 'E', 16: '5', 4: 'G', 22: '~', 15: '6', 17: '6', 29: '2', 0: 'k', 13: '4', 38: 'h', 33: '0', 26: '.', 36: 'U', 6: 'l'}
```

**Comment by yan12125** [comment]
Does not work. Sometimes it throws 403 and sometimes it returns wrong results:
```
$ youtube-dl -v test:openload                 
[debug] System config: []
[debug] User config: []
[debug] Custom config: []
[debug] Command-line args: ['-v', 'test:openload']
[debug] Encodings: locale UTF-8, fs utf-8, out UTF-8, pref UTF-8
[debug] youtube-dl version 2017.02.07
[debug] Git HEAD: 013877298
[debug] Python version 3.6.0 - Linux-4.9.6-1-ARCH-x86_64-with-arch
[debug] exe versions: ffmpeg 3.2.2, ffprobe 3.2.2
[debug] Proxy map: {}
[TestURL] Test URL: https://openload.co/f/kUEfGclsU9o
[Openload] kUEfGclsU9o: Downloading webpage
[debug] Invoking downloader on 'https://openload.co/stream/.0045o46t~Kv.116Y1EfGm1U8c021U73Jw4kB~l9s~.'
ERROR: unable to download video data: HTTP Error 403: Forbidden
Traceback (most recent call last):
  File "/home/yen/Projects/youtube-dl/youtube_dl/YoutubeDL.py", line 1704, in process_info
    success = dl(filename, info_dict)
  File "/home/yen/Projects/youtube-dl/youtube_dl/YoutubeDL.py", line 1646, in dl
    return fd.download(name, info)
  File "/home/yen/Projects/youtube-dl/youtube_dl/downloader/common.py", line 353, in download
    return self.real_download(filename, info_dict)
  File "/home/yen/Projects/youtube-dl/youtube_dl/downloader/http.py", line 61, in real_download
    data = self.ydl.urlopen(request)
  File "/home/yen/Projects/youtube-dl/youtube_dl/YoutubeDL.py", line 2006, in urlopen
    return self._opener.open(req, timeout=self._socket_timeout)
  File "/usr/lib/python3.6/urllib/request.py", line 532, in open
    response = meth(req, response)
  File "/usr/lib/python3.6/urllib/request.py", line 642, in http_response
    'http', request, response, code, msg, hdrs)
  File "/usr/lib/python3.6/urllib/request.py", line 570, in error
    return self._call_chain(*args)
  File "/usr/lib/python3.6/urllib/request.py", line 504, in _call_chain
    result = func(*args)
  File "/usr/lib/python3.6/urllib/request.py", line 650, in http_error_default
    raise HTTPError(req.full_url, code, msg, hdrs, fp)
urllib.error.HTTPError: HTTP Error 403: Forbidden


[1]$ youtube-dl -v test:openload
[debug] System config: []
[debug] User config: []
[debug] Custom config: []
[debug] Command-line args: ['-v', 'test:openload']
[debug] Encodings: locale UTF-8, fs utf-8, out UTF-8, pref UTF-8
[debug] youtube-dl version 2017.02.07
[debug] Git HEAD: 013877298
[debug] Python version 3.6.0 - Linux-4.9.6-1-ARCH-x86_64-with-arch
[debug] exe versions: ffmpeg 3.2.2, ffprobe 3.2.2
[debug] Proxy map: {}
[TestURL] Test URL: https://openload.co/f/kUEfGclsU9o
[Openload] kUEfGclsU9o: Downloading webpage
[debug] Invoking downloader on 'https://openload.co/stream/U6o51Uv02ar4f16Z4.c9~s010~l174.J3mq8kj~E6.G'
[download] Destination: skyrim_no-audio_1080.mp4-kUEfGclsU9o.mp4
[download] 100% of 31.00B in 00:00

$ cat skyrim_no-audio_1080.mp4-kUEfGclsU9o.mp4 
{"status":403,"msg":"wrong IP"}
```

**Comment by sudovijay** [comment]
@yan12125  This code works perfectly fine for me ! as i said i ain't a python guy ! 

Here try yourself !

`
ol_id = '1912600068121040814522121400721607315067330652607609130410762808802070250653206719095380702907317071131340712337065301180513039121031453407418075140682307336104010682109835145110682406827068201270613010067311244209004';

first_two_chars = int(float(ol_id[0:][:2]))
urlcode = {}
num = 2

while num < len(ol_id):
    key = int(float(ol_id[num + 3:][:2]))
    urlcode[key] = chr(int(float(ol_id[num:][:3])) - first_two_chars)
    num += 5

sorted(urlcode, key=lambda key: urlcode[key])  

urllink = ''.join(['%s' % (value) for (key, value) in urlcode.items()])

video_url = 'https://openload.co/stream/' + urllink

print(video_url)`

**Comment by yan12125** [comment]
Alright, apparently you didn't test on Python 3.
```
$ python2 test.py
https://openload.co/stream/kUEfGclsU9o~1486567011~113.193.0.0~O6hLofoi

$ python3 test.py
https://openload.co/stream/k1U~f560.9o9E3.0L364sh.cof~7816U1O~111lo0iG
```
This PR works for Python 2 only.

**Comment by sudovijay** [comment]
@yan12125  ! thats weird i just tried on both it works pretty fine ! 

```
$ python test.py 
https://openload.co/stream/kUEfGclsU9o~1486567011~113.193.0.0~O6hLofoi
$ python3 test.py
https://openload.co/stream/kUEfGclsU9o~1486567011~113.193.0.0~O6hLofoi
```
Even tried here too . it works fine

https://repl.it/languages/python3

**Comment by yan12125** [comment]
Maybe you've found a bug in Python 3.6. I'll have to investigate

**Comment by yan12125** [comment]
Your implementation relies on the order of ```dict``` objects. See this example:
```
d = {}

d['b'] = 2
d['a'] = 1
for k, v in d.items():
    print(k)
```
The result is:
```
$ python3.6 test2.py
b
a
$ python2.7 test2.py
a
b
```
The order of generic dictionaries is undefined in Python, so it may change in different Python versions. You'll need to rewrite this code so that it works on Python 3.6.

FYI:

1. ```sorted``` function does not change ```urlcode```. The input data is [copied](https://hg.python.org/cpython/file/2.7/Python/bltinmodule.c#l2239), sorted and then returned. You can remove this line and see what happens
2. In Python 3.6 the order of ```dict``` items are preserved. That is, the order of items() is the same as how you inserted elements. See http://bugs.python.org/issue27350 and https://docs.python.org/3/whatsnew/3.6.html#new-dict-implementation

**Comment by mdoulabi** [comment]
when update youtube-dl compile for this error ?

File: youtube_dl/extractor/openload.py (hunk 1 above)
Comment (cemeyer) [inline youtube_dl/extractor/openload.py]: Probably this will work instead?

`urllink = ''.join([str(urlcode[key]) for key in sorted(urlcode.keys())])`

**Comment by ozcs** [comment]
i just received a new update on the youtube-dl but the openload still not working , 

this is the error i m getting and running on python 2.7 , ubuntu 16.04

ValueError: unichr() arg not in range(0x110000) (wide Python build)

if there is no issue on using the openload with python3 , then plz advice how can i force youtube-dl to use python3 instead of normal python

**Comment by cemeyer** [comment]
Y'all, please use `urllink = ''.join([str(urlcode[key]) for key in sorted(urlcode.keys())])`.  It doesn't rely upon dictionary order.

**Comment by ozcs** [comment]
which line we need to replace?

**Comment by mdoulabi** [comment]
i change , but error : 
youtube-dl -v https://openload.co/embed/ZD36I7512Mw/
Traceback (most recent call last):
  File "/usr/lib64/python2.6/runpy.py", line 122, in _run_module_as_main
    "__main__", fname, loader, pkg_name)
  File "/usr/lib64/python2.6/runpy.py", line 34, in _run_code
    exec code in run_globals
  File "/usr/local/bin/youtube-dl/__main__.py", line 16, in <module>
  File "/usr/local/bin/youtube-dl/youtube_dl/__init__.py", line 43, in <module>
  File "/usr/local/bin/youtube-dl/youtube_dl/extractor/__init__.py", line 9, in <module>
  File "/usr/local/bin/youtube-dl/youtube_dl/extractor/extractors.py", line 54, in <module>
  File "/usr/local/bin/youtube-dl/youtube_dl/extractor/ard.py", line 7, in <module>
  File "/usr/local/bin/youtube-dl/youtube_dl/extractor/generic.py", line 85, in <module>
  File "/usr/local/bin/youtube-dl/youtube_dl/extractor/openload.py", line 107
SyntaxError: 'return' outside function

**Comment by yan12125** [comment]
@mdoulabi just update to the latest version.

**Comment by ozcs** [comment]
thanks , this fix is working for me now

**Comment by ozcs** [comment]
hi, i m just reporting that openload has stopped working again , 


[TestURL] Test URL: https://openload.co/f/kUEfGclsU9o
[Openload] kUEfGclsU9o: Downloading webpage
Traceback (most recent call last):
  File "/usr/local/bin/youtube-dl", line 11, in <module>
    sys.exit(main())
  File "/usr/local/lib/python2.7/dist-packages/youtube_dl/__init__.py", line 457, in main
    _real_main(argv)
  File "/usr/local/lib/python2.7/dist-packages/youtube_dl/__init__.py", line 447, in _real_main
    retcode = ydl.download(all_urls)
  File "/usr/local/lib/python2.7/dist-packages/youtube_dl/YoutubeDL.py", line 1883, in download
    url, force_generic_extractor=self.params.get('force_generic_extractor', False))
  File "/usr/local/lib/python2.7/dist-packages/youtube_dl/YoutubeDL.py", line 772, in extract_info
    return self.process_ie_result(ie_result, download, extra_info)
  File "/usr/local/lib/python2.7/dist-packages/youtube_dl/YoutubeDL.py", line 833, in process_ie_result
    extra_info=extra_info)
  File "/usr/local/lib/python2.7/dist-packages/youtube_dl/YoutubeDL.py", line 761, in extract_info
    ie_result = ie.extract(url)
  File "/usr/local/lib/python2.7/dist-packages/youtube_dl/extractor/common.py", line 427, in extract
    ie_result = self._real_extract(url)
  File "/usr/local/lib/python2.7/dist-packages/youtube_dl/extractor/openload.py", line 89, in _real_extract
    urlcode.append((key, compat_chr(int(ol_id[num + 2:num + 5]) // int(ol_id[num + 1]) - first_char)))
ValueError: invalid literal for int() with base 10: 'e92'

openload ID extraction error

**Comment by yan12125** [comment]
@ozcs Fixed in #12357

**Pull request closed without merging**